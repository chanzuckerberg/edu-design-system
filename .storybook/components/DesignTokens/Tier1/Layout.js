import React, {Component} from 'react';
import {Grid, GridItem, Section} from '../../../../src';
import filterTokens from '../../../util/filterTokens';
import {TokenSpecimen} from '../../TokenSpecimen/TokenSpecimen';

export class Tier1Layout extends Component {
  render() {
    return (
      <div>
        <Section title="Layout">
          <Grid>
            {filterTokens(`eds-l-`).map(function (listItem, index) {
              return (
                <GridItem key={'l-' + index}>
                  <TokenSpecimen
                    inlineStyles={{
                      maxWidth: `var(${listItem.name})`,
                    }}
                    name={listItem.name}
                    value={listItem.value}
                  />
                </GridItem>
              );
            })}
          </Grid>
        </Section>
      </div>
    );
  }
}
