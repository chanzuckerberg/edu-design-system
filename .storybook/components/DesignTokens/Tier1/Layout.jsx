import React, { Component } from 'react';

import filterTokens from '../../../util/filterTokens';
import { Grid } from '../../Grid/Grid';
import { Section } from '../../Section/Section';
import { TokenSpecimen } from '../../TokenSpecimen/TokenSpecimen';

export class Tier1Layout extends Component {
  render() {
    return (
      <div>
        <Section title="Layout">
          <Grid>
            {filterTokens(`eds-l-`).map(function (listItem) {
              return (
                <Grid.Item key={listItem.name}>
                  <TokenSpecimen
                    inlineStyles={{
                      maxWidth: `var(${listItem.name})`,
                    }}
                    name={listItem.name}
                    value={listItem.value}
                  />
                </Grid.Item>
              );
            })}
          </Grid>
        </Section>
      </div>
    );
  }
}
