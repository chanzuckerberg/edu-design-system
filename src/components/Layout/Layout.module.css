@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    # LAYOUT
\*------------------------------------*/

/**
 * 1) Component that controls an overarching page layout. By default,
 *    the layout renders a fixed-position left sidebar on larger screens
 * 2) Left sidebar is the default
 * 3) minmax(0, 1fr) used to prevent content from overflowing the viewport
 * when items are stacked
 */
.layout {
  display: grid;
  gap: var(--eds-size-3);
  grid-template-columns: minmax(0, 1fr); /* 1 */

  @media all and (min-width: $eds-bp-xl) {
    grid-template-columns: auto 1fr;
  }
}

/**
 * Layout right sidebar variant
 */
.layout--right-sidebar {
  @media all and (min-width: $eds-bp-xl) {
    grid-template-columns: 1fr auto;
  }
}

/**
 * Layout 67-33 variant
 */
.layout--67-33 {
  @media all and (min-width: $eds-bp-xl) {
    grid-template-columns: 2fr 1fr;
  }
}

/**
 * Layout no gap variant
 */
.layout--gap-none {
  gap: 0;
}

/**
 * Layout lg-xl gap variant
 */
.layout--gap-lg-xl {
  gap: var(--eds-size-5);
  @media all and (min-width: $eds-bp-xl) {
    gap: var(--eds-size-12);
  }
}

/**
 * Layout 50-50 variant with two main columns of equal width
 */
.layout--50-50 {
  @media all and (min-width: $eds-bp-xl) {
    grid-template-columns: 1fr 1fr;
  }
}

/**
 * Sidebar layout section region
 * 1) Contains the global header and site navigation
 */
.layout__section--sidebar {
  @media all and (min-width: $eds-bp-xl) {
    top: 0;
    left: 0;
    height: 100vh;
    width: var(--eds-l-sidebar-width);
    z-index: var(--eds-z-index-500);
  }

  /**
 * Layout 67-33 variant
 */
  .layout--67-33 & {
    @media all and (min-width: $eds-bp-xl) {
      width: 100%;
    }
  }
}

/**
 * Sidebar layout section region within right sidebar
 */
.layout--right-sidebar > .layout__section--sidebar {
  order: 2;
}

/**
 * Sidebar layout section region within wide sidebar variant
 * 1) Width should be larger than the default
 */
.layout--sidebar-wide > .layout__section--sidebar {
  @media all and (min-width: $eds-bp-xl) {
    width: 22rem; /* TODO: Make more dynamic */
  }
}

/**
 * Main layout section region 
 * 1) Contains the main page content
 */
.layout__section--main {
  @media all and (min-width: $eds-bp-xl) {
    width: 100%;
    height: 100vh;
  }
}

/**
 * Main layout section region within right sidebar
 * 1) Place below the right sidebar on small screens
 * 2) Place to the left of the sidebar on large screens
 */
.layout--right-sidebar > .layout__section--main {
  order: 1; /* 2 */
}
