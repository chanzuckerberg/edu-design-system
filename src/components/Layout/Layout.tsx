import clsx from 'clsx';
import React, { ReactNode } from 'react';
import styles from './Layout.module.css';

export interface Props {
  /**
   * Behavior
   */
  behavior?: 'fixed-sidebar';

  /**
   * Child node(s) that can be nested inside component
   */
  children: ReactNode;
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Expandable layout sections
   * 1) Used for hover/focus states showing/hiding extra content
   */
  expandable?: boolean;
  /**
   * Sidebar property
   * 1) Adjust the gap between layout sections.
   */
  gap?: 'none' | 'lg-xl';
  /**
   * Sidebar property
   * 1) Adjust the size of the sidebar
   */
  sidebar?: 'wide';
  /**
   * Sidebar property
   * 1) Adjust the size of the sidebar
   */
  variant?: 'right-sidebar' | '50-50' | '60-40';
}

/**
 * BETA: This component is still a work in progress and is subject to change.
 *
 * ```ts
 * import {Layout} from "@chanzuckerberg/eds";
 * ```
 *
 * Component that controls an overarching page layout. By default, the layout renders
 * a fixed-position left sidebar on larger screens.
 */
export const Layout = ({
  behavior,
  children,
  className,
  variant,
  gap,
  sidebar,
  ...other
}: Props) => {
  const componentClassName = clsx(
    styles['layout'],
    className,
    variant === '60-40' && styles['layout--60-40'],
    variant === 'right-sidebar' && styles['layout--right-sidebar'],
    variant && styles[`layout--${variant}`],
    behavior === 'fixed-sidebar' && styles['layout--fixed-sidebar'],
    gap && styles[`layout--gap-${gap}`],
    sidebar === 'wide' && styles['layout--sidebar-wide'],
  );

  return (
    <div className={componentClassName} {...other}>
      {children}
    </div>
  );
};
