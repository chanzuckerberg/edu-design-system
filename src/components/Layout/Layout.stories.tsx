import type {StoryObj, Meta} from '@storybook/react';
import React from 'react';

import {Layout} from './Layout';
import LayoutSection from '../LayoutSection';

export default {
  title: 'Molecules/Layout and Containers/Layout',
  component: Layout,
  args: {
    children: (
      <>
        <LayoutSection region="sidebar">
          <div className="fpo">Sidebar Layout Section</div>
        </LayoutSection>
        <LayoutSection region="main">
          <div className="fpo">Main Layout Section</div>
        </LayoutSection>
      </>
    ),
  },
} as Meta<Args>;

type Args = React.ComponentProps<typeof Layout>;

export const Default: StoryObj<Args> = {};

export const GapNone: StoryObj<Args> = {
  args: {
    gap: 'none',
  },
};

export const GapLgXl: StoryObj<Args> = {
  args: {
    gap: 'lg-xl',
  },
};

export const SixtySevenThirtyThree: StoryObj<Args> = {
  args: {
    children: (
      <>
        <LayoutSection region="main">
          <div className="fpo">Main Layout Section</div>
        </LayoutSection>
        <LayoutSection region="sidebar">
          <div className="fpo">Sidebar Layout Section</div>
        </LayoutSection>
      </>
    ),
    variant: '67-33',
  },
};

export const FiftyFifty: StoryObj<Args> = {
  args: {
    children: (
      <>
        <LayoutSection region="main">
          <div className="fpo">Main Layout Section 1</div>
        </LayoutSection>
        <LayoutSection region="main">
          <div className="fpo">Main Layout Section 2</div>
        </LayoutSection>
      </>
    ),
    variant: '50-50',
  },
};

export const ThirtyThreeSixtySeven: StoryObj<Args> = {
  args: {
    children: (
      <>
        <LayoutSection region="sidebar">
          <div className="fpo">Sidebar Layout Section</div>
        </LayoutSection>
        <LayoutSection region="main">
          <div className="fpo">Main Layout Section</div>
        </LayoutSection>
      </>
    ),
    variant: '33-67',
  },
};
