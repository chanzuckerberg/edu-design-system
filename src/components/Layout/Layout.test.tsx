import { generateSnapshots } from '@chanzuckerberg/story-utils';
import * as LayoutStoryFile from './Layout.stories';

describe('<Layout />', () => {
  generateSnapshots(LayoutStoryFile);
});
