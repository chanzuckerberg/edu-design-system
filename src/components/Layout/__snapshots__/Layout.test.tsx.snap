// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> Default story renders snapshot 1`] = `
<div
  class="layout"
>
  <div
    class="layout__section layout__section--sidebar"
  >
    <div
      class="fpo"
    >
      Sidebar Layout Section
    </div>
  </div>
  <div
    class="layout__section layout__section--main"
  >
    <div
      class="fpo"
    >
      Main Layout Section
    </div>
  </div>
</div>
`;

exports[`<Layout /> FiftyFifty story renders snapshot 1`] = `
<div
  class="layout layout--50-50"
>
  <div
    class="layout__section layout__section--main"
  >
    <div
      class="fpo"
    >
      Main Layout Section 1
    </div>
  </div>
  <div
    class="layout__section layout__section--main"
  >
    <div
      class="fpo"
    >
      Main Layout Section 2
    </div>
  </div>
</div>
`;

exports[`<Layout /> GapLgXl story renders snapshot 1`] = `
<div
  class="layout layout--gap-lg-xl"
>
  <div
    class="layout__section layout__section--sidebar"
  >
    <div
      class="fpo"
    >
      Sidebar Layout Section
    </div>
  </div>
  <div
    class="layout__section layout__section--main"
  >
    <div
      class="fpo"
    >
      Main Layout Section
    </div>
  </div>
</div>
`;

exports[`<Layout /> GapNone story renders snapshot 1`] = `
<div
  class="layout layout--gap-none"
>
  <div
    class="layout__section layout__section--sidebar"
  >
    <div
      class="fpo"
    >
      Sidebar Layout Section
    </div>
  </div>
  <div
    class="layout__section layout__section--main"
  >
    <div
      class="fpo"
    >
      Main Layout Section
    </div>
  </div>
</div>
`;

exports[`<Layout /> RightSidebar story renders snapshot 1`] = `
<div
  class="layout layout--right-sidebar"
>
  <div
    class="layout__section layout__section--main"
  >
    <div
      class="fpo"
    >
      Main Layout Section
    </div>
  </div>
  <div
    class="layout__section layout__section--sidebar"
  >
    <div
      class="fpo"
    >
      Sidebar Layout Section
    </div>
  </div>
</div>
`;

exports[`<Layout /> SidebarWide story renders snapshot 1`] = `
<div
  class="layout layout--sidebar-wide"
>
  <div
    class="layout__section layout__section--sidebar"
  >
    <div
      class="fpo"
    >
      Sidebar Layout Section
    </div>
  </div>
  <div
    class="layout__section layout__section--main"
  >
    <div
      class="fpo"
    >
      Main Layout Section
    </div>
  </div>
</div>
`;
